Tests: Replace k_current_get() in ISR context

kokas-a opened the pull request:
This fix repairs https://github.com/zephyrproject-rtos/zephyr/issues/62705 and https://github.com/zephyrproject-rtos/zephyr/issues/62704 that caused by incorrect using k_current_get() in ISR context. The routine was replaced with k_sched_current_thread_query().

Commit:
Tests: Replace k_current_get() in ISR context

This fix repairs https://github.com/zephyrproject-rtos/zephyr/issues/62705 and
https://github.com/zephyrproject-rtos/zephyr/issues/62704
that caused by incorrect using k_current_get() in ISR context.
The routine was replaced with k_sched_current_thread_query().

Signed-off-by: Nikolay Agishev <agishev@synopsys.com>

kokas-a commented:
@ruuddw , could you please take a look?

ruuddw commented:
@kokas-a thanks, this looks like the correct fix to me.
Fixing the tests instead of trying to somehow make TLS work in ISR/exception context.

andyross reviewed (commented):
These should only collide when TLS+userspace is enabled[1], it's got nothing to do with ISR state per se, unless I'm missing something.

What's breaking that leads you to suspect you need this?

[1]  In other circumstances you just pull the thread pointer out of the current CPU struct.

andyross commented:
To elaborate: k_current_get() has worked in this circumstance for longer than I've been working on Zephyr.  This is one of **the** most stable of stable APIs.  If we broke it we should fix it, not go around modifying tests.

andyross reviewed (changes requested):
Heh, and having elaborated like that... I think this is now a -1 from me.  We seem to have broken a core kernel API and we can't do that.  Can you file a bug instead?  @dcpleung or another TLS expert: what's the underlying issue here?

dcpleung commented:
I think the fundamental issue is that... TLS is supposed to not be valid in interrupt context as it is not executing code as a thread. And `k_current_get()` grabs thread pointer via TLS by default if TLS is enabled, and thus cannot really be used inside ISRs. This works for now because most architectures does not modify the "TLS pointer" during interrupt entry so it still points to the running thread just prior to interrupt entry.

andyross commented:
Yeah, that's deeper than this patch.  Again, `k_current_get()`  is the correct API for getting the current thread pointer, and it's worked in ISRs for years and years and years.  If it doesn't work in ISRs now, that's because we broke it.  That's a bug in k_current_get(), and the solution definitely isn't to add a new API everyone is supposed to use instead.

Can't we just put a `k_is_in_isr()` check in it and indirect to the expensive version #if TLS, etc...?

npitre commented:
My $0.02:

The linked bugs are 1) related to TLS and 2) only on ARC.

So my recommendation would be for ARC to either 1) make sure the TLS base
pointer is valid even in interrupt context (there is no fundamental reason 
why this can't be) or 2) disable the TLS variation in the definition of
`k_current_get()` for ARC.

ruuddw commented:
I don't think ARC TLS is broken or should be fixed. I think the tests are broken. 

1. Is it legal to call `k_current_get` from ISR context? What should the result be: error/invalid, the task running when the ISR occurred, new ready task that will be resumed when returning from ISR? What does it mean in SMP context?
My suggestions/answers: either undefined result (should not be called), or error

2.  Is it legal to assume TLS can be used in ISR context? Should “ISR context” have its own TLS instance?
My suggestion: no valid TLS in ISR context

3. Same questions for exception handler context (the second test calls` k_current_get()` from a fatal error handler hook.

npitre commented:
Good points. However extra considerations are needed.

About (3): Fatal error handlers definitely have legitimate `k_current_get()` 
use cases. When a thread crashes, the error handler must be able to display
which thread it was executing in its crash dump. Sometimes it might be a
recoverable error depending on some thread-specific setting/permission, etc.

About (1): This is more subtle but `k_current_get()` may be useful with
profilers. The most useful result is obviously the task running when the 
ISR occurred on the local CPU. This is less useful on SMP if you can't
control timer IRQ affinity but on !SMP this is perfectly usable.

About (2): Unless Zephyr implements threaded IRQ handlers it doesn't make
much sense to have TLS instances for IRQ context.

Now Andy is right about `k_current_get()` having worked in ISRs for years.
What's new is the (relatively) recent optimization where the current thread
pointer is stuffed in the TLS area. This apparently works just fine on
all supported architectures but ARC. Should we deny this functionality to
all because ARC doesn't cope with that optimization? I don't think that is
the right compromize.

So the logical choice is again for ARC to skip the current-in-TLS 
optimization when in IRQ/exception context, or ensure the dedicated CPU 
register holding a pointer to the TLS area remains valid in IRQ context 
(that register is otherwise unused by the compiler anyway).

ruuddw commented:
The reason the TLS base pointer is not (easily) available on certain ARC configurations is that they use multiple register banks. One bank is dedicated for interrupt processing. Super fast interrupt handling, saves context saving time.
Syncing the TLS register in both register banks every context switch is extra context switch overhead. And disabling the TLS-cached also has performance impact.
But if the agreed semantics for k_current_get is to provide the id of the last thread that ran (and not necessarily the 'current' one), ARC will have to conform.
BTW, what's are the cases where the 'query' variant is intended to be used?

npitre commented:
> BTW, what's are the cases where the 'query' variant is intended to be used?

Mostly when TLS content is not yet initialized i.e. during early boot
before `arch_tls_stack_setup()` has been called.

All this being said...

I'm having second thoughts. When user space is used, the TLS lives in user 
memory. We surely don't want unprivileged threads pretending to be some 
other thread by altering the TLS content and fool privileged code via 
syscalls... or IRQ handlers.

In that regard, most of the privileged code, like the scheduler, uses 
`_current` instead which, incidentally, is defined as 
`k_sched_current_thread_query()`. But not all the privileged code does that.

I'm therefore wondering if the current-in-TLS is such a good idea in the end.

ceolin commented:
> Yeah, that's deeper than this patch. Again, `k_current_get()` is the correct API for getting the current thread pointer, and it's worked in ISRs for years and years and years. If it doesn't work in ISRs now, that's because we broke it. That's a bug in k_current_get(), and the solution definitely isn't to add a new API everyone is supposed to use instead.

Agree ... IMHO TLS shouln't be used at all. I hit this problem in Xtensa and end up doing it https://github.com/zephyrproject-rtos/zephyr/pull/61303/commits/0b616a6332a618802f45060bdc6d48229413fa32 

> 
> Can't we just put a `k_is_in_isr()` check in it and indirect to the expensive version #if TLS, etc...?

I got the impression of using TLS here is to avoid a syscall in user mode, if we need to call `k_is_in_isr()` we should just make it a syscall and use `__current`

ceolin commented:
> My $0.02:
> 
> The linked bugs are 1) related to TLS and 2) only on ARC.
> 
> So my recommendation would be for ARC to either 1) make sure the TLS base pointer is valid even in interrupt context (there is no fundamental reason why this can't be) 

Even if you have a valid TLS base pointer this can be missinterpreted, one thread  setting a timer `k_timer_init(&my_timer, expire_fn, stop_fn)` my think that TLS variables are from that thread ...

or 2) disable the TLS variation in the definition of `k_current_get()` for ARC.

ruuddw commented:
I hadn't considered the security impact yet, but indeed having privileged code use a spoofed "current thread" sounds bad. E.g. fatal error handler killing the wrong thread.
And, it looks like xtensa has similar issues with assuming TLS to be valid in ISR context.

Should the logic be reversed, and the syscall to get _current be the default `k_current_get()`, with an optional "k_current_from_TLS" or so instead of the `k_sched_current_thread_query()` as the exception?

ruuddw commented:
All, can we accept this PR to fix the ARC test/CI issues with TLS enabled?
I still think calling `k_sched_current_thread_query()` in ISR/fatal context is currently the best solution, in line with the API definitions (use 'query' if TLS not available/working).
Reverting the caching of current in TLS or rethinking/discussing that approach can be a new, more generic issue/PR.

andyross commented:
> All, can we accept this PR to fix the ARC test/CI issues with TLS enabled?

But again, the test code isn't wrong.  Can't you just file a bug against k_current_get()+TLS on ARC and filter out the test on the affected platforms instead?

Commit:
ARC: workaroud for TLS-related tests

This PR makes workaround for
https://github.com/zephyrproject-rtos/zephyr/issues/62705 and
https://github.com/zephyrproject-rtos/zephyr/issues/62704

In current discussion it was mentioned that it's incorrect to get
current running thread and its attributes from exception handler
with k_current_get() as additional register banks may appear
unsincronized.

Signed-off-by: Nikolay Agishev <agishev@synopsys.com>

andyross commented:
> Changes in the PR are specific to ARC

No, it's still changing the test code, which isn't incorrect.  Again, I understand that this API is hard to implement correctly on ARC than TLS is in use.  The test is still repsonsible for testing that the platform implements the API correctly, which it doesn't.  If you can't fix it, then choose one of the options above for how to work around it (filter the test case, disallow TLS under whatever circumstance, document the shortcoming in a bug that you'll leave open, etc...).

You absolutely can't hide a test failure by changing the test.  It's a real test failure!

ruuddw commented:
Hi @andyross, by completely disabling or filtering the test we loose all other parts of the test that should be tested also when TLS is enabled. This way we prevent that. My understanding is that instead of using the alternative implementation that does work (and is correct according the API documentation to use if the other doesn't...), you prefer to don't do somehow disable the specific test for ARC+TLS? I don't accept fully skipping the tests, so it would be #ifdef's at some other place.

andyross commented on tests/kernel/context/src/main.c:
```diff
@@ -127,7 +127,7 @@ static void isr_handler(const void *data)
 	switch (isr_info.command) {
 	case THREAD_SELF_CMD:
-		isr_info.data = (void *)k_current_get();
+		isr_info.data = (void *)k_sched_current_thread_query();
```
Just to repeat: this here is what we can't do.  k_current_get() is the **correct API** for the test to use here.  It's not working, and that's a bug that the test has discovered.   If we're not going to fix the bug (and it sounds like you're not), you need to document that limitation somewhere and just accept that this test case won't pass.

And you don't need to filter the whole of tests/kernel/context.  Just a a `ztest_test_skip()` (when ARC && TLS) to the cases that hit this code in ISR context.

kokas-a replied:
Hi @andyross,
Regarding your advice I tried to cut off ```context``` and ```context_one_cpu``` tests for ARC.
To make compiller happy, I had to apply a lot of #ifdef conditions inside the code, that make the code ugly:
https://github.com/zephyrproject-rtos/zephyr/commit/cfc363619212a120b5e634bd12f7171f0efd6739

Another option - is apply runtime construction:
```
	if(IS_ENABLED(NO_TLS_IN_EXCEPTION)){
	    ztest_test_skip();
	}else{
	    ...
	}
```

Which one option is more applicable for current situation?

ruuddw commented:
To keep things simple, I'm fine with the runtime check instead of compile time. Acceptable for you as well @andyross ?

kokas-a commented:
Issues fixed in https://github.com/zephyrproject-rtos/zephyr/pull/67831 with ARC-only related changes.
